### ISSUE TITLE
New dashboard plot discussion

### ISSUE THREAD

**Issue by jacobtomlinson** [dask/distributed#3318]
I've been experimenting with a new dashboard element to help visualise transfer between workers and overall cluster communication. This is a super early rough mockup of the kind of thing we could have. I'd like to share it and have some discussion around it before I put more time into it.

Currently it shows tasks being sent to workers and memory transfer between workers.

Things on my list of improvements so far are:
- Place workers spatially based on their network performance. Workers with higher bandwidth to each other should be closer together.
- Workers could show their memory and CPU usages.
- Clients could be shown and client-scheduler transfers could be shown. (We could have the client notify the scheduler of incoming work if it is about to serialise a lot of things, the apparent hang here is a common grumble).
- Various UCX protocols could be shown somehow instead of just the red transfer lines.
- Show disk-read and disk-write operations somehow.

Feedback/criticism would be appreciated!

![ezgif com-optimize(2)](https://user-images.githubusercontent.com/1610850/70787442-902e5480-1d86-11ea-8e7a-a0b742dca499.gif)

**Comment by mrocklin** [comment]
> Place workers spatially based on their network performance. Workers with higher bandwidth to each other should be closer together.

FWIW I would personally find the networking structure to be more valuable.  For example I'd love to see all of the workers grouped by hostname.  Our measurement of bandwidth is often somewhat messy and not necessarily particularly telling.

> Clients could be shown and client-scheduler transfers could be shown. (We could have the client notify the scheduler of incoming work if it is about to serialise a lot of things, the apparent hang here is a common grumble).

+1

> Various UCX protocols could be shown somehow instead of just the red transfer lines.

For the common-case Dask user UCX use is still somewhat uncommon.  Dask also may not have good visibility into what was used.  Instead, I would encourage us to focus on the amount of transfer (maybe with duration) and the bandwidth of transfer (maybe with width), which are more universally accessible and important.

> Show disk-read and disk-write operations somehow.

It might be useful for workers to publish their disk usage as a metric to be stored back up in the scheduler.  This could be a small incremental PR if you want to get your feet wet with pushing telemtry back up to the scheduler.

**Comment by mrocklin** [comment]
Also, thanks for doing this, I'm enthusiastic to see what eventually comes out of this

**Comment by jacobtomlinson** [comment]
Thanks Matt!

> For example I'd love to see all of the workers grouped by hostname.

I agree this would be useful. I'm also conscious of scenarios where workers may be on the same machine but not necessarily aware of that because they are in containers. So bandwidth could be a useful proxy for this.

I'm also curious about how machine locality on cloud platforms could be shown. Are workers in the same rack, the same datacenter, the same region? Does the number of hops make a big difference? Can we map this out somehow. Again bandwitdth could be a useful proxy rather than implementing infrastructure specific stuff.

> I would encourage us to focus on the amount of transfer (maybe with duration) and the bandwidth of transfer (maybe with width), which are more universally accessible and important.

Yes fair enough. @quasiben suggested line widths to show bandwidth of transfers too.

> It might be useful for workers to publish their disk usage as a metric to be stored back up in the scheduler. This could be a small incremental PR if you want to get your feet wet with pushing telemtry back up to the scheduler.

Sounds fun!

**Comment by jacobtomlinson** [comment]
Thanks Matt!

> For example I'd love to see all of the workers grouped by hostname.

I agree this would be useful. I'm also conscious of scenarios where workers may be on the same machine but not necessarily aware of that because they are in containers. So bandwidth could be a useful proxy for this.

I'm also curious about how machine locality on cloud platforms could be shown. Are workers in the same rack, the same datacenter, the same region? Does the number of hops make a big difference? Can we map this out somehow. Again bandwitdth could be a useful proxy rather than implementing infrastructure specific stuff.

> I would encourage us to focus on the amount of transfer (maybe with duration) and the bandwidth of transfer (maybe with width), which are more universally accessible and important.

Yes fair enough. @quasiben suggested line widths to show bandwidth of transfers too.

> It might be useful for workers to publish their disk usage as a metric to be stored back up in the scheduler. This could be a small incremental PR if you want to get your feet wet with pushing telemtry back up to the scheduler.

Sounds fun!

**Comment by jacobtomlinson** [comment]
Now that we have a first version in I'm exploring how I can show more information about a worker. Currently, we have an orange circle which increases in size and changes to the task color when it is running a task.

Additionally, I think it would be useful to show the following information:
- Worker name
- CPU Usage
- Memory usage
- Number of cores/threads the worker has
- Whether it has an accelerator device like a GPU
- Whether it is experiencing connection problems

I would also like to know whether it would be useful to communicate whether it is holding futures and how many.

Here's a quick draft of what a worker could look like on the dashboard with these enhancements.

![image](https://user-images.githubusercontent.com/1610850/72612927-0e51be00-3926-11ea-9663-28f88db60313.png)

- The colored circles indicate CPU/memory
- The orange dots indicate the number of cores/threads
- The warning triangle would be shown in the case of issues
- The PCI device icon would be shown when an accelerator is being used

I would love some feedback on this design before I move forward to implementation.

**Comment by mrocklin** [comment]
I personally wasn't familiar with the PCI device icon.  That might need some explanation?

We might want to include both the address and the name, or maybe even just the name with a hover for the address.

**Comment by mrocklin** [comment]
So the annuli are the CPU/Memory bits?  It seems a little cluttered at first glance.  I wonder what would happen if you removed the black border, and just had a gap of white space instead.  That moves the apparent number of annuli from 3 to 2, which might help.

**Comment by jacobtomlinson** [comment]
> I personally wasn't familiar with the PCI device icon.

Fair enough. Is there some other way to communicate that the worker has additional hardware resources like a GPU which may be more familiar?

> We might want to include both the address and the name, or maybe even just the name with a hover for the address.

I'm tempted to hide this all together and then show both on hover.

> So the annuli are the CPU/Memory bits?

Yes

>  It seems a little cluttered at first glance.

That's reasonable. Perhaps hiding most of this except the worker and annuli and just showing it on hover would be better?

**Comment by mrocklin** [comment]
Maybe.  I think that we need to use it a bit and start asking what questions we're looking for when we consult the dashboard.  

Questions like "How is my memory doing?", "How active are my GPUs?", and "How many workers do I have and how many cores does each have like reasonable questions to want answered quickly.  I don't know yet how we'll use this plot though, and if some of those questions will be better served by other plots.  I'm inclined to get something in so that people start using it, and then ask around for feedback before getting very creative here.

**Comment by jacobtomlinson** [comment]
After putting more thought to this I feel like this dashboard is going to be of most use to get a qualitative understanding of your cluster.

The main question I want to answer is:

- "What is my cluster doing right now?"

This dashboard provides a 1000 foot view of the cluster to answer initial questions such as:

- "How many workers do I have?"
- "What resources do those workers have?"
- "How well utilized are their resources?"
- "Are there currently any issues in the cluster?"
- "Are workers spending most of their time computing or communicating?"
- "Which protocols are being used to communicate?"
- "How much memory has been swapped to disk?"

It not only should answer questions like these but answer them all at once to give a high-level overview.

Then if a user identifies an issue or something they want to understand better they can drill into the more quantitative plots and performance reports.

Therefore I feel like attempting to communicate as many different aspects as possible, but each in as subtle/minimal a way as possible would provide the most value here. Then providing a flow into more detailed diagnostics would be useful. Perhaps even having a click-through from various elements to other dashboards (although that makes less sense in the lab extension).

**Comment by quasiben** [comment]
Thanks @jacobtomlinson for that break down of questions a dashboard like this could answer.  And while dask does have detailed informative pages, building a summary page with a lot of info subtly communicated would be very useful.  Often it seems like users are putting this "page" together with the dask labextension, resource viz below,  task viz plot on top, etc  -- organized the way a user can see everything all at once.

Would you collect other summary information as well ?  Max/min/avg of tasks, communications etc ?  

The beginner question you noted I would say is very valuable.  I'm interpreting it as, I just created a cluster -- did I do it correctly ?  This comes up fairly often even for very experienced dask users

**Comment by jacobtomlinson** [comment]
> Would you collect other summary information as well ? Max/min/avg of tasks, communications etc ?

Sure.

> I just created a cluster -- did I do it correctly ? This comes up fairly often even for very experienced dask users

Yup I certainly ask this question a lot. Especially with cloud cluster managers, or GPUs or alternative transports. Or in times where Dask is extremely helpful and makes things for your (creating `Client()` with no other context for example), it would be helpful to see what happened.

**Comment by mrocklin** [comment]
> "What is my cluster doing right now?"
> Therefore I feel like attempting to communicate as many different aspects as possible, but each in as subtle/minimal a way as possible would provide the most value here

I'm on board with that.  I'm excited to see you come back to this.
